middle_end/flambda/simplify/simplify_common.ml

```diff
@@ -43,13 +43,43 @@ type 'a expr_simplifier =
 type simplify_toplevel =
      Downwards_acc.t
   -> Expr.t
+  -> code_id:Code_id.t option
   -> return_continuation:Continuation.t
   -> return_arity:Flambda_arity.With_subkinds.t
   -> Exn_continuation.t
   -> return_cont_scope:Scope.t
   -> exn_cont_scope:Scope.t
   -> Expr.t * Upwards_acc.t
 
+let is_self_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.function_code_id denv with
+  | Some fun_code_id ->
+    begin match Apply.call_kind apply with
+    | Function (Direct { code_id = apply_code_id; _ }) ->
+      Code_id.equal fun_code_id apply_code_id
+    | _ -> false
+    end
+  | None -> false
+
+let is_tail_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.function_continuation denv,
+        DE.function_exn_continuation denv with
+  | None, None -> false
+  | Some fun_cont, Some fun_exn_cont ->
+    begin match Apply.continuation apply with
+    | Return apply_cont ->
+      let apply_exn_cont = Apply.exn_continuation apply in
```

Comment:
I added a check to enforce that only functions are considered as tail-calls, because we may be interested in knowing whether a call is a tailcall independently of whether it is a self-tailcall, e.g. tailcalls in a set of mutually recursive functions.